### PR TITLE
feat: magic-link; password-free login

### PR DESCRIPTION
**Pull request by larsivi**
This is an addon service to password and adds the ability to authenticate via a token. The token can be sent to you via email using the requestLoginToken endpoint, or by other means server side by generating a token and adding it using addLoginToken. Users can now be found with this token, using findUserByLoginToken.
Default expiry is 15 minutes.

Three new packages are added, magic-link and mongo-magic-link and client-magic-link.

There is a logical dependency on package that handles users (typically password which in any case is a good companion module), but it is not hard if you have a different one providing the same.

Example of use (serverside, accompanied by password):
```ts
// There is nothing specific in Database setup, except the need for the mongo-magic-link package

const accountsPassword = new AccountsPassword({});
const accountsMagicLink = new AccountsMagicLink({});
const accountsServer = new AccountsServer(
       {
            db: accountsDb,
            tokenSecret: "terrible secret"
        },
        {
            password: accountsPassword,
            magicLink: accountsMagicLink,
        }
);
 ```

Example of use clientside (using GraphQL):
```ts
const apolloClient = new ApolloClient({ ...apolloOptions });
const accountsGraphQL = new GraphQLClient({graphQLClient: apolloClient});
accountsClient = new AccountsClient({}, accountsGraphQL);
const accountsMagicLink = new AccountsClientMagicLink(accountsClient);

 // Requesting email
await accountsMagicLink.requestMagicLinkEmail(email);

// Then, in the login handler
const loginResult = await accountsMagicLink.login({ token });
```

### CODE REVIEW

File: packages/token/src/accounts-token.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+  public isTokenExpired(tokenRecord: TokenRecord, expiryDate: number): boolean {
+    return Number(tokenRecord.when) + expiryDate < Date.now();
+  }
+
+  private async tokenAuthenticator(token: string): Promise<CustomUser> {
+    const foundUser: CustomUser | null = await this.db.findUserByLoginToken(token);
+
+    if (!foundUser) {
+      if (this.server.options.ambiguousErrorMessages) {
```
Comment (pradel) [inline packages/token/src/accounts-token.ts:110]: I am not sure this is really needed as we authenticate with the login token and not an email here

**Review by pradel — COMMENTED** [review]
Thank you very much for the pr, this will be a great addition that a lot of people requested!

As a general note, what do you think about renaming `@accounts/token` to `@accounts/magic-link`? The name is pretty standard for these kinds of authentication services, so it will be clear what this package is doing.

Regarding the logic, the implementation sounds good, I didn't do a deep review for this one but left a few small comments for now.
I was not able to compile the packages locally, graphql-client is throwing an error.

Security questions:
- when requesting 2 emails, should both be active or just the last one sent? So when I ask for a new email it invalidates all the previous tokens, only the last email link is working.

**Comment by pradel** [comment]
More comments, documentation and examples are still missing for this, is this something you would be up to do?

**Comment by larsivi** [comment]
> More comments: documentation and examples are still missing for this, is this something you would be up to do?

Yes, but if possible in a later PR.

File: packages/client-token/src/client-token.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+   */
+  public async login(user: LoginUserTokenService): Promise<LoginResult> {
+    return this.client.loginWithService('token', user);
+  }
+
+  /**
+   * Request a new login token.
+   * @param {string} email - The email address to send a password reset link.
+   */
+  public requestLoginToken(email: string): Promise<void> {
```
[the remark this answers is not in the document]
Comment (larsivi) [reply packages/client-token/src/client-token.ts:25]: I expect us to need this for SMS, which is why I went for a more generic variant, but I see your point (and I think you said you didn't want SMS support in accounts).

File: website/docs/strategies/password.md [hunk 3]
```diff
@@ -88,7 +88,7 @@ The password module come with two factor out of the box. You can customize it us
 Check all the options available [here](/docs/api/two-factor/interfaces/accountstwofactoroptions).
 
 ```javascript
-export const accountsPassword = new AccountsPassword({
+export const accountsPassword = new AccountsToken({
```
[the remark this answers is not in the document]
Comment (larsivi) [reply website/docs/strategies/password.md:91]: You're right.

File: packages/token/src/accounts-token.ts (hunk 1 above)
Comment (larsivi) [reply packages/token/src/accounts-token.ts:110]: Exactly what isn't needed? Maybe I am misunderstanding something.

**Comment by larsivi** [comment]
> As a general note, what do you think about renaming `@accounts/token` to `@accounts/magic-link`? The name is pretty standard for these kinds of authentication services, so it will be clear what this package is doing.

Sounds good!

> Security questions:
> 
> * when requesting 2 emails, should both be active or just the last one sent? So when I ask for a new email it invalidates all the previous tokens, only the last email link is working.
> * after a successful login all the tokens should be invalidated

Both of these makes sense, but maybe they could/should be covered by options? Or tied to the reason, such that login tokens explicitly added through addLoginToken could be managed by other means.

File: packages/client-magic-link/__tests__/client-magic-link.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  loginWithService: jest.fn(),
+};
+const accountsMagicLink = new AccountsClientMagicLink(mockedClient as any);
+
+const user = {
+  email: 'john@doe.com',
+  token: 'deadbeef',
+};
+
+describe('AccountsClientPassword', () => {
```
Comment (pradel) [inline packages/client-magic-link/__tests__/client-magic-link.ts:16]:
```suggestion
describe('AccountsClientMagicLink', () => {
```

File: packages/client-magic-link/package.json [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  "files": [
+    "src",
+    "lib"
+  ],
+  "jest": {
+    "preset": "ts-jest"
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/accounts-js/accounts/tree/master/packages/client-token"
```
Comment (pradel) [inline packages/client-magic-link/package.json:31]:
```suggestion
    "url": "https://github.com/accounts-js/accounts/tree/master/packages/client-magic-link"
```

File: packages/client-magic-link/src/client-magic-link.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  /**
+   * Log the user in with a token.
+   */
+  public async login(user: LoginUserMagicLinkService): Promise<LoginResult> {
+    return this.client.loginWithService('magic-link', user);
+  }
+
+  /**
+   * Request a new login token.
+   * @param {string} email - The email address to send a password reset link.
```
Comment (pradel) [inline packages/client-magic-link/src/client-magic-link.ts:23]:
```suggestion
   * @param {string} email - The email address to send a login link.
```

File: packages/database-manager/__tests__/database-manager.ts [hunk 7]
```diff
@@ -243,4 +255,16 @@ describe('DatabaseManager', () => {
   it('removeAllResetPasswordTokens should be called on userStorage', () => {
     expect(databaseManager.removeAllResetPasswordTokens('userId')).toBe('userStorage');
   });
+
+  it('addLoginToken should be called on sessionStorage', () => {
```
Comment (pradel) [inline packages/database-manager/__tests__/database-manager.ts:259]:
```suggestion
  it('addLoginToken should be called on userStorage', () => {
```

File: packages/database-mongo-magic-link/src/mongo-magic-link.ts [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    this.password = this.options.password;
+  }
+
+  /**
+   * Setup the mongo indexes needed for the token service.
+   * @param options Options passed to the mongo native `createIndex` method.
+   */
+  public async setupIndexes(options: Omit<IndexOptions, 'unique' | 'sparse'> = {}): Promise<void> {
+    // Token index used to verify the email address of a user
+    await this.userCollection.createIndex('services.token.loginTokens.token', {
```
Comment (pradel) [inline packages/database-mongo-magic-link/src/mongo-magic-link.ts:97]: should we change token to magicLik here?

File: website/docs/transports/graphql.md [hunk 9]
```diff
@@ -201,7 +201,7 @@ By default accounts-js only allow 'username', 'email' and 'password' for the use
 ```javascript
 // server.js
 
-const accountsPassword = new AccountsPassword({
+const accountsPassword = new AccountsMagicLink({
```
Comment (pradel) [inline website/docs/transports/graphql.md:204]: please revert

File: packages/types/src/types/services/magic-link/database-interface.ts [hunk 10]
```diff
@@ -0,0 +1,9 @@
+import { User } from '../../user';
+
+export interface DatabaseInterfaceServiceMagicLink<CustomUser extends User = User> {
+  findUserByLoginToken(token: string): Promise<CustomUser | null>;
+
+  addLoginToken(userId: string, email: string, token: string, reason: string): Promise<void>;
```
Comment (pradel) [inline packages/types/src/types/services/magic-link/database-interface.ts:6]: do we really need a reason here?

**Review by pradel — COMMENTED** [review]
- can you provide an example of the user creation? My understanding is that a password would be required in the current state? I might be missing something obvious tho :D

- I guess you deleted the `yarn.lock` file and it was regenerated from scratch, this is creating some issues, could you please revert the change you made and update the one from master instead of recreating it?

File: packages/magic-link/src/accounts-magic-link.ts [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+  }
+
+  public isTokenExpired(tokenRecord: TokenRecord, expiryDate: number): boolean {
+    return Number(tokenRecord.when) + expiryDate < Date.now();
+  }
+
+  private async magicLinkAuthenticator(token: string): Promise<CustomUser> {
+    const foundUser: CustomUser | null = await this.db.findUserByLoginToken(token);
+
+    if (!foundUser) {
```
Comment (pradel) [inline packages/magic-link/src/accounts-magic-link.ts:115]: if user is not found it means that token is invalid, better to throw an `incorrectToken` error

Comment (pradel) [reply packages/magic-link/src/accounts-magic-link.ts:115]: also, can we rename `incorrectToken ` to `invalidToken` for consistency with password service?

Comment (pradel) [reply packages/magic-link/src/accounts-magic-link.ts:115]: actually I was taking a look at the password service and we should throw `LoginTokenExpired `. That way, an attacker can't guess if the token is expired or does not exist.

File: packages/magic-link/src/accounts-magic-link.ts [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+        this.options.errors.userNotFound,
+        RequestMagicLinkEmailErrors.UserNotFound
+      );
+    }
+
+    // Remove pre-existing login tokens on user
+    await this.db.removeAllLoginTokens(user.id);
+
+    const token = generateRandomToken();
+    await this.db.addLoginToken(user.id, email, token, 'magic-link-requested');
```
Comment (pradel) [inline packages/magic-link/src/accounts-magic-link.ts:74]: why do we need a reason here?

File: packages/database-mongo-magic-link/src/mongo-magic-link.ts (hunk 8 above)
Comment (larsivi) [reply packages/database-mongo-magic-link/src/mongo-magic-link.ts:97]: The service name should definately change to magicLink - was that what you meant?

File: packages/types/src/types/services/magic-link/database-interface.ts (hunk 10 above)
Comment (larsivi) [reply packages/types/src/types/services/magic-link/database-interface.ts:6]: No, I am removing it.

File: packages/magic-link/src/accounts-magic-link.ts (hunk 12 above)
Comment (larsivi) [reply packages/magic-link/src/accounts-magic-link.ts:74]: I have removed the reason param

File: packages/magic-link/src/accounts-magic-link.ts (hunk 11 above)
Comment (larsivi) [reply packages/magic-link/src/accounts-magic-link.ts:115]: No ambigiousErrors handling? Always LoginTokenExpired? Would that also apply to the token record check itself?

Comment (pradel) [reply packages/magic-link/src/accounts-magic-link.ts:115]: Yeah I think so, to be consistent with https://github.com/accounts-js/accounts/blob/master/packages/password/src/accounts-password.ts#L307

**Comment by pradel** [comment]
@larsivi 

```
// There is nothing specific in Database setup, except the need for the mongo-magic-link package

const accountsPassword = new AccountsPassword({});
const accountsMagicLink = new AccountsMagicLink({});
const accountsServer = new AccountsServer(
       {
            db: accountsDb,
            tokenSecret: "terrible secret"
        },
        {
            password: accountsPassword,
            magicLink: accountsMagicLink,
        }
);

```

regarding this, it's a bit weird that you have to add the password package too, this means that all the mutations (or routes via REST) will be exposed and user creation via password will be allowed.
Would it be possible to just install the magic link one?

**Comment by larsivi** [comment]
> regarding this, it's a bit weird that you have to add the password package too, this means that all the mutations (or routes via REST) will be exposed and user creation via password will be allowed.
> Would it be possible to just install the magic link one?

Well, with the last cleanup I think that should work. The example was mostly like that since I do use the password module for user creation (and also for alternative login method). I haven't tried without though.

**Comment by pradel** [comment]
Okay, would be great to have the example so we are sure it's working that way. Will make it easier for users to get started with magic-link!

**Comment by acomito** [comment]
you guys are gods

**Comment by pradel** [comment]
Amazing work @larsivi thank you so much for this!

**Comment by pradel** [comment]
Everything is published on npm, would be cool to get some documentation for other users!

**Comment by larsivi** [comment]
Thank you! I'll look at writing some docs over the next few weeks.

**Pull request merged**